File: 0680-Valid Palindrome II/0680-Valid Palindrome II.cpp
class Solution {
public:
    bool validPalindrome(string s) {
        if (s.size() == 0) {
            return true;
        }
        
        int start = 0, end = s.size() - 1;
        while (start < end) {
            if (s[start] != s[end]) {
                return isPalindrome(s.substr(start, end - start)) || isPalindrome(s.substr(start + 1, end - start));
            }
            ++start;
            --end;
        }
        
        return true;
    }
    
private:
    bool isPalindrome(string s) {
        if (s.size() == 0) {
            return true;
        }
        
        int start = 0, end = s.size() - 1;
        while (start < end) {
            if (s[start] != s[end]) {
                return false;
            }
            ++start;
            --end;
        }
        
        return true;
    }
};

File: 0601-0700/0680-Valid Palindrome II/0680-Valid Palindrome II.py
class Solution:
    def validPalindrome(self, s: str) -> bool:
        def isPalindrome(i, j):
            return all(s[k] == s[j - k + i] for k in range(i, (i + j) // 2 + 1))
        
        for i in range(len(s) // 2):
            if s[i] != s[~i]:
                j = len(s) - 1 - i
                return  isPalindrome(i+1, j) or isPalindrome(i, j - 1)
        return True
